### PR TITLE
Attempt to unify queue state

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
* Mostly deprecate queue ack level in favor of queue state
* Additional compatibility logic for forward / backward compatibility
* Use queue state for replication acks
* Allow replication streams between clusters with different shard count

**Why?**
N/A

**How did you test it?**
N/A

**Potential risks**
N/A

**Is hotfix candidate?**
N/A

### CODE REVIEW

**Comment by wxing1292** [comment]
depends on #4176

**Comment by wxing1292** [comment]
additional UT in progress

File: service/history/shard/compatibility.go [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+func loadShardInfoCompatibilityCheck(
+	clusterMetadata cluster.Metadata,
+	shardInfo *persistencespb.ShardInfo,
+) *persistencespb.ShardInfo {
+	// TODO this section maintains the forward / backward compatibility
+	//  should be removed once the migration is done
+	//  also see ShardInfoToBlob
+
+	allClusterInfo := clusterMetadata.GetAllClusterInfo()
+	shardInfo = loadShardInfoCompatibilityCheckWithoutReplication(shardInfo)
```
Comment (yycptt) [inline service/history/shard/compatibility.go:46]: I think it's ok to get rid of this (and the store counterpart), all queues except replication have already migrated to use queueStates.

File: service/history/queues/queue_base_test.go [hunk 2]
```diff
@@ -731,3 +732,20 @@ func (s *queueBaseSuite) TestUpdateReaderProgress() {
...
 }
+
+func (s *queueBaseSuite) QueueStateEqual(
+	this *persistencespb.QueueState,
+	that *persistencespb.QueueState,
+) {
+	thisBlob, err := serialization.QueueStateToBlob(this)
+	s.NoError(err)
+	this, err = serialization.QueueStateFromBlob(thisBlob.Data, thisBlob.EncodingType.String())
+	s.NoError(err)
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/queues/queue_base_test.go:743]: sure, this is dealing with the time equal w/ timezone issue

File: service/history/shard/context_impl.go [hunk 3]
```diff
@@ -2059,6 +1933,13 @@ func copyShardInfo(shardInfo *persistencespb.ShardInfo) *persistencespb.ShardInf
 			ClusterAckLevel: maps.Clone(ackLevels.ClusterAckLevel),
 		}
 	}
+	// need to ser/de to make a deep copy of queue state
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/shard/context_impl.go:1936]: e.g. tests passing a shard info and change that shard info causing weird issue

File: service/history/replication/task_processor_manager.go [hunk 4]
```diff
@@ -315,7 +322,44 @@ func (r *taskProcessorManagerImpl) cleanupReplicationTasks() error {
...
+		) {
+			readerIDs = append(readerIDs, shard.ReplicationReaderIDFromClusterShardID(
+				targetClusterID, targetShardID,
+			))
+		}
+	}
+	return readerIDs
+}
+
+func minInt64(
```
Comment (yycptt) [inline service/history/replication/task_processor_manager.go:357]: `common/util.Min[int64]()`

File: service/history/shard/compatibility.go [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+		return shardInfo
+	}
+
+	for readerID, readerState := range queueStates.ReaderStates {
+		clusterID, _ := ReplicationReaderIDToClusterShardID(readerID)
+		clusterName, _, _ := ClusterNameInfoFromClusterID(allClusterInfo, clusterID)
+		queueAckLevel, ok := shardInfo.QueueAckLevels[tasks.CategoryIDReplication]
+		if !ok {
+			queueAckLevel = &persistencespb.QueueAckLevel{
+				AckLevel:        0,
```
Comment (yycptt) [inline service/history/shard/compatibility.go:209]: just double checking: replication doesn't use this?

File: service/history/shard/compatibility.go [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+				delete(shardInfo.QueueStates[tasks.CategoryIDReplication].ReaderStates, readerID)
+			}
+		}
+	}
+	return shardInfo
+}
+
+// ReplicationReaderIDFromClusterShardID convert from cluster ID & shard ID to reader ID
+// NOTE: cluster metadata guarantee
+//  1. initial failover version <= int32 max
```
Comment (yycptt) [inline service/history/shard/compatibility.go:255]: how do we guarantee this? I think existing definition uses int64?

File: service/history/shard/context_impl.go [hunk 7]
```diff
@@ -1246,63 +1142,64 @@ func (s *ContextImpl) updateShardInfoLocked() error {
...
-	visibilityLag := s.immediateTaskExclusiveMaxReadLevel - s.getQueueAckLevelLocked(tasks.CategoryVisibility).TaskID - 1
-	timerLag := s.timeSource.Now().Sub(s.getQueueAckLevelLocked(tasks.CategoryTimer).FireTime)
-
-	handler.Histogram(metrics.ShardInfoReplicationLagHistogram.GetMetricName(), metrics.ShardInfoReplicationLagHistogram.GetMetricUnit()).Record(replicationLag)
-	handler.Histogram(metrics.ShardInfoTransferLagHistogram.GetMetricName(), metrics.ShardInfoTransferLagHistogram.GetMetricUnit()).Record(transferLag)
-	handler.Histogram(metrics.ShardInfoVisibilityLagHistogram.GetMetricName(), metrics.ShardInfoVisibilityLagHistogram.GetMetricUnit()).Record(visibilityLag)
-	handler.Timer(metrics.ShardInfoTimerLagTimer.GetMetricName()).Record(timerLag)
-}
+//
+//func (s *ContextImpl) emitShardInfoMetricsLogsLocked() {
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/shard/context_impl.go:1146]: this actually reminded me that the metrics is not fixed yet. 
will update

File: service/history/shard/compatibility.go (hunk 5 above)
Comment (wxing1292) [reply service/history/shard/compatibility.go:209]: i do not believe replication use ack level var (this should be used by transfer Q, time Q, etc)
replication acks are per cluster, so only `ClusterAckLevel map[string]int64` should be used

File: service/history/shard/compatibility.go (hunk 6 above)
Comment (wxing1292) [reply service/history/shard/compatibility.go:255]: the cluster.Metadata init logic <s>does</s> additional check .. ok i forgot to copy the logic from my stretch pad
```zsh
temporal|queue-state-replication ⇒ git show queue-state-replication-old -- common/cluster 
───────┬─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ STDIN
───────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ commit 5faf42979c395dc521ac9835805ed7b52c3ba20c (queue-state-replication-old)
   2   │ Author: wenquan xing <wenquan.xing@gmail.com>
   3   │ Date:   Tue Apr 18 16:09:58 2023 -0700
   4   │ 
   5   │     Using queue state for replication stream
   6   │ 
   7   │ diff --git a/common/cluster/metadata.go b/common/cluster/metadata.go
   8   │ index 4cabedc68..a4f33d097 100644
   9   │ --- a/common/cluster/metadata.go
  10   │ +++ b/common/cluster/metadata.go
  11   │ @@ -29,6 +29,7 @@ package cluster
  12   │  import (
  13   │     "context"
  14   │     "fmt"
  15   │ +   "math"
  16   │     "strconv"
  17   │     "sync"
  18   │     "sync/atomic"
  19   │ @@ -158,8 +159,8 @@ func NewMetadata(
  20   │         panic("Master cluster name is empty")
  21   │     } else if len(currentClusterName) == 0 {
  22   │         panic("Current cluster name is empty")
  23   │ -   } else if failoverVersionIncrement == 0 {
  24   │ -       panic("Version increment is 0")
  25   │ +   } else if failoverVersionIncrement <= 0 || failoverVersionIncrement > math.MaxInt32 {
  26   │ +       panic("Version increment <= 0 or > 2147483647")
  27   │     }
  28   │  
  29   │     versionToClusterName := updateVersionToClusterName(clusterInfo, failoverVersionIncrement)
───────┴─────────────────────────────────────────────────────────────────────────────────────────────────
```

File: service/history/shard/context_util.go [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+func getMinTaskKey(
+	queueState *persistencespb.QueueState,
+) *tasks.Key {
+	var minTaskKey *tasks.Key
+	if queueState.ExclusiveReaderHighWatermark != nil {
+		taskKey := ConvertFromPersistenceTaskKey(queueState.ExclusiveReaderHighWatermark)
+		minTaskKey = &taskKey
+	}
+	for _, readerState := range queueState.ReaderStates {
+		taskKey := ConvertFromPersistenceTaskKey(readerState.Scopes[0].Range.InclusiveMin)
```
Comment (yycptt) [inline service/history/shard/context_util.go:158]: I see a couple of places where [0] is directly access, I do understand right now scopes must be non-empty, but someone may break it in the future... :( 

I am thinking if is it possible to trim those empty slices from map on loading shard info & Set/UpdateQueueState? Maybe you have a better idea?

**Review by yycptt — APPROVED** [review]

File: service/history/shard/context_util.go (hunk 8 above)
Comment (wxing1292) [reply service/history/shard/context_util.go:158]: sure, will send another PR

Comment (wxing1292) [reply service/history/shard/context_util.go:158]: actually, does https://github.com/temporalio/temporal/pull/4190/files/167dc0be71db8e4010720c04f8a3021133fa01b4#diff-14665987ac1f72fe1eec59b480b6463c209c117352308819c88585805ec66fc8R227 address your concern?

Comment (yycptt) [reply service/history/shard/context_util.go:158]: this only performs trimming for replication? I think we'll want to protect all categories.

**Pull request merged**